react/components/organisms/form/form.js

```diff
@@ -0,0 +1,248 @@
+    static get propTypes() {
+        return {
+            fields: PropTypes.arrayOf(
+                PropTypes.shape({
+                    title: PropTypes.string,
+                    fields: PropTypes.array.isRequired
+                })
+            ).isRequired,
+            values: PropTypes.object.isRequired,
+            rejectButtonProps: PropTypes.object,
+            acceptButtonProps: PropTypes.object,
+            saveNotification: PropTypes.bool,
+            saveMessage: PropTypes.func,
+            error: PropTypes.bool,
+            errorMessage: PropTypes.func,
+            onUpdateValues: PropTypes.func,
+            onDiscard: PropTypes.func,
+            onSave: PropTypes.func,
+            style: ViewPropTypes.style
+        };
+    }
+
+    static get defaultProps() {
+        return {
+            rejectButtonProps: {},
+            acceptButtonProps: {},
+            saveNotification: true,
+            saveMessage: values => "Changes saved!",
+            error: true,
+            errorMessage: error => error.message || "Something went wrong",
+            onUpdateValues: values => {},
+            onDiscard: undefined,
+            onSave: undefined,
+            style: {}
+        };
+    }
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            valuesData: this.props.values,
+            saving: false
+        };
+
+        this.firstInput = null;
+    }
+
+    componentDidMount() {
+        this.firstInput?.focus();
+    }
+
+    save = async () => {
+        if (!this.props.onSave) return;
+        this.setState({ saving: true });
+
+        try {
+            await this.props.onSave(this.state.valuesData);
+
+            if (this.saveNotification) {
+                notify(this.props.saveMessage(this.state.valuesData));
+            }
+        } catch (error) {
+            if (this.error) {
+                this.notify(this.props.errorMessage(error));
+            }
+        } finally {
+            this.setState({ saving: false });
+        }
+    };
+
+    discard = async () => {
+        if (!this.props.onDiscard) return;
+
+        try {
+            await this.props.onDiscard(this.state.valuesData);
+        } catch (error) {
+            if (this.props.error) {
+                this.notify(this.props.errorMessage(error));
+            }
+        }
+    };
+
+    onValueUpdate = (name, value) => {
+        this.setState(
+            prevState => ({
+                valuesData: { ...prevState.valuesData, [name]: value }
+            }),
+            () => this.props.onUpdateValues(this.state.valuesData)
+        );
+    };
+
+    onSubmit = async () => {
+        await this.save();
+    };
+
+    onReject = async () => {
+        await this.discard();
+    };
+
+    _style = () => {
+        return [styles.form, this.props.style];
+    };
+
+    _renderInputComponent = ({ value, type, label, ...args }) => {
+        switch (type) {

```

Comment:
I dont know if this is applicable, but if this form could be used to edit a password field, it would be nice to have it hidden while you type (like `"***4"). Perhaps that could be another type?

Reply:
It could be passed in the input args

Reply:
Right, nevermind. Nice